hudi-client/src/main/java/org/apache/hudi/table/action/commit/UpsertPartitioner.java

```diff
@@ -272,21 +275,44 @@ public int getPartition(Object key) {
       String partitionPath = keyLocation._1().getPartitionPath();
       List<InsertBucket> targetBuckets = partitionPathToInsertBuckets.get(partitionPath);
       // pick the target bucket to use based on the weights.
-      double totalWeight = 0.0;
       final long totalInserts = Math.max(1, profile.getWorkloadStat(partitionPath).getNumInserts());
       final long hashOfKey = NumericUtils.getMessageDigestHash("MD5", keyLocation._1().getRecordKey());
       final double r = 1.0 * Math.floorMod(hashOfKey, totalInserts) / totalInserts;
-      for (InsertBucket insertBucket : targetBuckets) {
-        totalWeight += insertBucket.weight;
-        if (r <= totalWeight) {
-          return insertBucket.bucketNumber;
-        }
+
+      int index = binarySearch(targetBuckets, r);
+      if (index >= 0) {
+        return targetBuckets.get(index).bucketNumber;
+      }
+
+      if (-1 * index - 1 < targetBuckets.size()) {
```

Comment:
sorry why do we need this? if not found, why not return the first bucket ?

Reply:
also, ideally we should not hit this scenario only. If our bin packing is good, cumulative weight of last entry should be 1.0.

Reply:
> sorry why do we need this? if not found, why not return the first bucket ?

Collections.binarySearch() will return the index of the search key, if it is contained in the list;   otherwise, (-(insertion point) - 1).  The insertion point is defined as the point at which the key would be inserted into the list: the index of the first element greater than the key, or list.size() if all  elements in the list are less than the specified key.

Reply:
gotcha. can we add braces. ((-1*index) -1) in both if condition and in the return statement

Reply:
Sure, I will fix the comments.